go-controller/pkg/node/linkmanager/link_network_manager.go

```diff
@@ -0,0 +1,389 @@
+	}
+	if address.IPNet.IP.IsUnspecified() {
+		return fmt.Errorf("IP must be specified")
+	}
+	link, err := netlinkOps.LinkByIndex(address.LinkIndex)
+	if err != nil {
+		return fmt.Errorf("no valid link associated with addresses %s: %v", address.String(), err)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// overwrite label to the name of this component in-order to aid address ownership. Label must start with link name.
+	address.Label = GetAssignedAddressLabel(link.Attrs().Name)
+	c.addAddressToStore(link.Attrs().Name, address)
+	c.reconcile()
+	return nil
+}
+
+// DelAddress removes the address from the store and ensure its removed from a link
+func (c *Controller) DelAddress(address netlink.Addr) error {
+	if address.LinkIndex == 0 {
+		return fmt.Errorf("link index must be non-zero")
+	}
+	if address.IPNet == nil {
+		return fmt.Errorf("IP must be non-nil")
+	}
+	if address.IPNet.IP.IsUnspecified() {
+		return fmt.Errorf("IP must be specified")
+	}
+	link, err := netlinkOps.LinkByIndex(address.LinkIndex)
+	if err != nil {
+		return fmt.Errorf("no valid link associated with addresses %s: %v", address.String(), err)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.delAddressFromStore(link.Attrs().Name, address)
+	c.reconcile()
+	return nil
+}
+
+func (c *Controller) reconcile() {
+	// 1. get all the links on the node
+	// 2. iterate over the links and get the addresses associated with it
+	// 3. cleanup any stale addresses from link that we no longer managed
+	// 4. remove any stale addresses from links that we do manage
+	// 5. add addresses that are missing from a link that we managed
+	links, err := netlinkOps.LinkList()
+	if err != nil {
+		klog.Errorf("Link Network Manager: failed to list links: %v", err)
+		return
+	}
+	for _, link := range links {
+		// get all addresses associated with the link depending on which IP families we support
+		addressesFound, err := getAllLinkAddressesByIPFamily(link, c.ipv4Enabled, c.ipv6Enabled)
+		if err != nil {
+			klog.Errorf("Link Network Manager: failed to get address from link %q", link.Attrs().Name)
+			continue
+		}
+		addressesWanted, found := c.store[link.Attrs().Name]
+		if !found {
+			// we dont managed this link, check we  used to manage it and if a stale address is present, delete it
+			for _, address := range addressesFound {
+				// we label any address we create, so if we aren't managed a link, we must remove any stale addresses
+				if address.Label == GetAssignedAddressLabel(link.Attrs().Name) {
+					if err := netlinkOps.AddrDel(link, &address); err != nil {
+						klog.Errorf("Link Network Manager: failed to delete address %q from link %q",
+							address.String(), link.Attrs().Name)
+					} else {
+						klog.Infof("Link Network Manager: successfully removed stale address %q from link %q",
+							address.String(), link.Attrs().Name)
+					}
+				}
+			}
+			continue
+		}
+		// ensure any addresses found that we previously managed are deleted
+		for _, addrFound := range addressesFound {
```

Comment:
this loop seems almost identical to what is running int he above loop with if !found. Can you just do the checks for if !found or if!containsAddress then do a single delete afterwards?

Reply:
done